rse-py-scripting.Rmd

```diff
@@ -103,42 +133,461 @@ optional arguments:
+
+We don't need to write any of this when we first create a new tool.
+However,
+experience teaches that little tools grow into big ones.
+As soon as we have more than a couple of configuration parameters,
+or as soon as some of them are usually set to the same value,
+we should start to think about saving our settings in files.
+
+This is particularly important if we want others (including our future self) to be able to reproduce our work.
+Careful record keeping is essential to reproducible science,
+and if we add just a few lines to our program,
+the computer can save the entire configuration for a particular run of a program.
+For example,
+if we have stored our settings in a Python dictionary called `settings` like this:
 
-Using this template as a starting point,
-we can add the functions we developed previously to a script called...
+```python
+settings = {
+    'logfile': '/tmp/thesis.log',
+    'quiet': False,
+    'overwrite': True,
+    'fonts': ['Comic Sans']
+}
+```
+
+we can save that as YAML using this short function:
+
+```python
+def save_config(filename, settings):
+    '''Save configuration settings to a file.'''
+
+    with open(filename, 'w') as writer:
+        yaml.dump(settings, writer)
+
+save_config('thesis.yml', settings)
+```
+
+```shell
+$ cat thesis.yml
+```
 
-FIXME: In this section we will take the code developed in the novice book
-and turn it into a Python script that can be executed at the command line.
-See [this lesson](https://carpentrieslab.github.io/python-aos-lesson/04-cmdline/index.html)
-for an example of this approach in action.
+```text
+logfile: "/tmp/thesis.log"
+quiet: false
+overwrite: true
+fonts:
+- Comic Sans
+```
 
-> **Code Reuse**
+This lets us re-create configuration on another machine
+even if it has different default settings.
+The test is whether our program can load a dumped configuration,
+then dump it again and get the same result.
+
+> **Version Everything**
 >
-> In the novice lessons,
-> we learned how to reuse (rather than cut and paste)
-> code by defining functions (Section \@ref(FIXME)).
-> In order to use those functions in other python notebooks/scripts,
-> we saw that we can save them in a file (called a module)
-> that can be imported (Section \@ref(FIXME)).
-> In this chapter, we've seen that we can go one step further
-> and run our Python code outside of a Python environment,
-> by writing Python scripts that can be executed at the command line.
+> We should always include a version number as a field in the dumped configuration
+> (Section \@ref(rse-git-advanced-tag)),
+> and our program should print this when given a `--version` option.
+> We need this because how we interpret options will change over time,
+> and if you don't know what the version of the program was,
+> we'll have to guess what options mean.
+
+## How can I detect errors? {#rse-py-scripting-exceptions}
+
```

Comment:
Perhaps error handing should have its own chapter of the book? (Perhaps the very next chapter after scripting?)

Reply:
I think it would be pretty short to be a chapter on its own. On the other hand, if the Python scripting chapter is re-labelled "How to write a robust command-line software tool", then maybe the config and error handling stuff fits more naturally?

Reply:
I guess the question that came to mind was "is error handling an explicitly command line tool thing, or are there other contexts in which it's applicable such as when working more interactively in a notebook?" We can leave it in this chapter for now :-)